crates/red_knot/tests/cli.rs

```diff
@@ -208,96 +172,159 @@ stat = add(10, 15)
 fn unknown_rules() -> anyhow::Result<()> {
-    let tempdir = TempDir::new()?;
-
-    let project_dir = tempdir.path().canonicalize()?;
-
-    std::fs::write(
-        project_dir.join("pyproject.toml"),
-        r#"
-[tool.knot.rules]
-division-by-zer = "warn" # incorrect rule name
-"#,
-    )
-    .context("Failed to write `pyproject.toml`")?;
-
-    std::fs::write(project_dir.join("test.py"), r#"print(10)"#)
-        .context("Failed to write `test.py`")?;
-
-    insta::with_settings!({filters => vec![(&*tempdir_filter(&project_dir), "<temp_dir>/")]}, {
-        assert_cmd_snapshot!(knot().current_dir(project_dir), @r"
-        success: false
-        exit_code: 1
-        ----- stdout -----
-        warning[unknown-rule] <temp_dir>/pyproject.toml:3:1 Unknown lint rule `division-by-zer`
-
-        ----- stderr -----
+    let case = TestCase::with_files([
+        (
+            "pyproject.toml",
+            r#"
+            [tool.knot.rules]
+            division-by-zer = "warn" # incorrect rule name
+            "#,
+        ),
+        ("test.py", "print(10)"),
+    ])?;
+
+    case.insta_settings().bind(|| {
+        assert_cmd_snapshot!(case.command(), @r"
+            success: false
+            exit_code: 1
+            ----- stdout -----
+            warning[unknown-rule] <temp_dir>/pyproject.toml:3:1 Unknown lint rule `division-by-zer`
+
+            ----- stderr -----
         ");
     });
 
     Ok(())
 }
 
-fn knot() -> Command {
-    Command::new(get_cargo_bin("red_knot"))
+struct TestCase {
+    _temp_dir: TempDir,
+    project_dir: PathBuf,
+}
+
+impl TestCase {
+    fn new() -> anyhow::Result<Self> {
+        let temp_dir = TempDir::new()?;
+
+        // Canonicalize the tempdir path because macos uses symlinks for tempdirs
+        // and that doesn't play well with our snapshot filtering.
+        let project_dir = temp_dir
+            .path()
+            .canonicalize()
+            .context("Failed to canonicalize project path")?;
+
+        Ok(Self {
+            project_dir,
+            _temp_dir: temp_dir,
+        })
+    }
+
+    fn with_files<'a>(files: impl IntoIterator<Item = (&'a str, &'a str)>) -> anyhow::Result<Self> {
+        let case = Self::new()?;
+        case.write_files(files)?;
+        Ok(case)
+    }
+
+    fn with_file(path: impl AsRef<Path>, content: &str) -> anyhow::Result<Self> {
+        let case = Self::new()?;
+        case.write_file(path, content)?;
+        Ok(case)
+    }
+
+    fn write_files<'a>(
+        &self,
+        files: impl IntoIterator<Item = (&'a str, &'a str)>,
+    ) -> anyhow::Result<()> {
+        for (path, content) in files {
+            self.write_file(path, content)?;
+        }
+
+        Ok(())
+    }
+
+    fn write_file(&self, path: impl AsRef<Path>, content: &str) -> anyhow::Result<()> {
+        let path = path.as_ref();
+
+        let path = if path.is_relative() {
+            self.project_dir.join(path)
+        } else {
+            path.to_path_buf()
```

Comment:
What would it mean for a test to write to an absolute path? Wouldn't that escape the temporary directory and potentially pollute/corrupt the developer's filesystem?

Reply:
I think it's fair to say that we should not support absolute paths for now.